### PR TITLE
Feature/category api

### PR DESCRIPTION
**Pull request by Darkosz1012**
Adds entire api to handle categories, with partial unit tests. So far I haven't added authorization on mutations.

### CODE REVIEW

File: __tests__/qraphql_queries/category.test.js [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+                        distance{
+                            name
+                        }
+                    }
+                }
+            `;
+
+            tester.test(true, query);
+        });
+        test("should failed if _if isn't specified.", () => {
```
Comment (frogi16) [inline __tests__/qraphql_queries/category.test.js:31]: Typo, should be "if _id"

File: __tests__/unit/graphql/loaders/createDefaultLoader.test.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    const competition_data = [
+        {
+            _id: '507f191e810c19729de860ea',
+            owner_id: "609aa4bde6483525a06b8e5b",
+            name: 'comp1',
+            start_date: "1989-12-31T23:00:00.000Z",
+            end_date: "1989-12-31T23:00:00.000Z",
+            location: 'sherwood',
+            details_id: "60a42ec1778fc82384125710",
+            categories_id:  []
```
Comment (frogi16) [inline __tests__/unit/graphql/loaders/createDefaultLoader.test.js:17]: Use single (') or double (") quotation marks, but be consistent.

File: __tests__/unit/graphql/loaders/createDefaultLoader.test.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    ];
+
+    beforeAll(async () => {
+        loader = createDefaultLoader(Competition)
+        mockingoose(Competition).toReturn(competition_data, "find");
+    });
+    test("Should return array with the correct number of elements.", async()=>{
+        const result = await loader.loadMany(["507f191e810c19729de860ea","507f191e810c19729de860ea","507f191e810c19729de860ed"]);
+
+        expect(result.length).toBe(3);
```
Comment (frogi16) [inline __tests__/unit/graphql/loaders/createDefaultLoader.test.js:38]: expect(result).toHaveLength(3)

File: graphql/resolvers/category/mutations.js [hunk 4]
```diff
@@ -1,19 +1,24 @@
-import { Category } from "$/db/index.js";
+import { Category, Competition,Distance, Score, Finals} from "$/db/index.js";
 
 export default {
     addCategory: async (parent, args, context, info) => {
-        let result = await Category.create(args.category);
+        let result = await Category.create(args);
+        await Competition.updateOne({_id:result.competition_id},{$push:{categories_id: result._id}});
```
Comment (frogi16) [inline graphql/resolvers/category/mutations.js:6]: Formatting is terrible, is you Prettier up and running?

File: graphql/resolvers/category/queries.js [hunk 5]
```diff
@@ -2,11 +2,15 @@ import { Category } from "$/db/index.js";
...
 export default {
     categories: async (parent, args, context, info) => {
-        let result = await Category.find();
+        let result = [];
+        if ("competition_id" in args) {
+            result = await Category.find({competition_id: args.competition_id});
+        }else{
+            result = await Category.find();
+        }
         return result;
```
Comment (frogi16) [inline graphql/resolvers/category/queries.js:11]:
```suggestion
        if ("competition_id" in args)
            return await Category.find({competition_id: args.competition_id});
        
        return await Category.find();
```

File: graphql/typedefs/mutation.graphql [hunk 6]
```diff
@@ -24,4 +21,21 @@ type Mutation {
...
+    ): Category
+    updateCategory(
+        _id: ID!
+        name: String
+        gender: String
+        start_stand: Int
+        end_stand: Int
+    ): Category
+    deleteCategory(_id: String!): Category
 }
```
Comment (frogi16) [inline graphql/typedefs/mutation.graphql:41]: Separate mutations with blank lines

**Review by frogi16 — CHANGES_REQUESTED** [review]

File: __tests__/qraphql_queries/category.test.js [hunk 7]
```diff
@@ -0,0 +1,258 @@
+import EasyGraphQLTester from "easygraphql-tester";
```
Comment (pieczonkam) [inline __tests__/qraphql_queries/category.test.js:1]: There are some minor problems that make several tests in this file fail (running these tests clearly says what's wrong with them, therefore I'm not gonna point out every single line that causes error). My guess is that newest commits provided changes that weren't taken into account in this test. Please, fix that.

**Review by pieczonkam — CHANGES_REQUESTED** [review]

File: __tests__/qraphql_queries/category.test.js (hunk 7 above)
Comment (Darkosz1012) [reply __tests__/qraphql_queries/category.test.js:1]: Great that you noticed that, I forgot to check the graphql tests after making changes to the schema.

File: __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js [hunk 8]
```diff
@@ -43,33 +43,20 @@ describe("createLoaderWithSpecificKey: dataloader with specific parameter for gr
...
-                categories_id: [],
-            },
+        const data = [
+            competition_data[0],
+            competition_data[1],
         ];
-        const result = await loader.loadMany(["609aa4bde6483525a06b8e5b"]);
-        expect(transform_by_JSON(result[0])).toMatchObject(competition_data);
+        const result = await loader.loadMany(["609aa4bde6483525a06b8e5b"])
+        console.log(result[1])
```
Comment (frogi16) [inline __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js:55]: Remove it

File: __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js [hunk 9]
```diff
@@ -43,33 +43,20 @@ describe("createLoaderWithSpecificKey: dataloader with specific parameter for gr
         mockingoose(Competition).toReturn(competition_data, "find");
     });
     test("Should return array with the correct number of elements.", async () => {
-        const result = await loader.loadMany(["609aa4bde6483525a06b8e5b"]);
-        expect(result).toHaveLength(1);
+        const result = await loader.loadMany(["609aa4bde6483525a06b8e5b","609aa4bde6483525a06b8e5z"]);
+        expect(result).toHaveLength(2);
     });
```
Comment (frogi16) [inline __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js]: Statements like "correct number of elements" are meaningless.
Try "should return array with object for every requested id" or something similar.

File: __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js [hunk 10]
```diff
@@ -43,33 +43,20 @@ describe("createLoaderWithSpecificKey: dataloader with specific parameter for gr
...
     test("Should return array with the correct structure of object.", async () => {
```
Comment (frogi16) [inline __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js:49]: Add blank line above

File: __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js [hunk 11]
```diff
@@ -43,33 +43,20 @@ describe("createLoaderWithSpecificKey: dataloader with specific parameter for gr
...
+            competition_data[0],
+            competition_data[1],
         ];
-        const result = await loader.loadMany(["609aa4bde6483525a06b8e5b"]);
-        expect(transform_by_JSON(result[0])).toMatchObject(competition_data);
+        const result = await loader.loadMany(["609aa4bde6483525a06b8e5b"])
+        console.log(result[1])
+        expect(transform_by_JSON(result[0])).toMatchObject(data);
+    });
+    test("Should return empty array when key when the key is not present in the data.", async () => {
```
Comment (frogi16) [inline __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js:58]: Add blank line

File: __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        const result = await loader.loadMany(["609aa4bde6483525a06b8e5b","609aa4bde6483525a06b8e5z"]);
+        expect(result).toHaveLength(2);
+    });
+    test("Should return array with the correct structure of object.", async () => {
+        const data = [
+            competition_data[0],
+            competition_data[1],
+        ];
+        const result = await loader.loadMany(["609aa4bde6483525a06b8e5b"])
+        console.log(result[1])
```
Comment (pieczonkam) [inline __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js:55]: Remove console.log before merging

File: __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js (hunk 11 above)
Comment (frogi16) [inline __tests__/unit/graphql/loaders/createLoaderWithSpecificKey.test.js:58]: Rename to "when key is not present in the data, corresponding object should be empty" to make more obvious that not whole result array should be empty.

**Review by frogi16 — CHANGES_REQUESTED** [review]

**Review by pieczonkam — APPROVED** [review]
Apply one small change, please. After that you can merge

**Review by frogi16 — APPROVED** [review]

**Pull request merged**